### PR TITLE
Salesforce add more tests

### PR DESCRIPTION
**Pull request by emregucerr**
## What does this PR do?

- Fixes #XXXX (GitHub issue number)
- Fixes CAL-XXXX (Linear issue number - should be visible at the bottom of the GitHub issue description)

## Mandatory Tasks (DO NOT REMOVE)

- [ ] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

- Are there environment variables that should be set?
- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)?
- Any other important info that could help to test that PR

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

### CODE REVIEW

File: packages/app-store/salesforce/lib/CrmService.test.ts [hunk 1]
```diff
@@ -1,12 +1,71 @@
...
 
 import type { CredentialPayload } from "@calcom/types/Credential";
 
+import type { appDataSchema } from "../zod";
 import SalesforceCRMService from "./CrmService";
 import { SalesforceRecordEnum } from "./enums";
 
+type AppOptions = z.infer<typeof appDataSchema>;
+
+const contactQueryResponse = {
```
Comment (fumeoss) [inline packages/app-store/salesforce/lib/CrmService.test.ts:14]: Consider extracting these mock responses into a shared test utilities file to avoid duplication and make them reusable across test files.

File: packages/app-store/salesforce/lib/CrmService.ts [hunk 2]
```diff
@@ -444,17 +444,15 @@ export default class SalesforceCRMService implements CRM {
...
-
-        const ownersQuery = await Promise.all(
-          Array.from(ownerIds).map(async (ownerId) => {
-            const result = await this.getSalesforceUserFromUserId(ownerId);
-            return result;
-          })
-        );
+        const ownersQuery =
+          ownerIds.size > 0
+            ? await Promise.all(
```
Comment (fumeoss) [inline packages/app-store/salesforce/lib/CrmService.ts:449]: Consider adding error handling to the Promise.all for owner queries to handle potential failures gracefully. Currently, if any owner query fails, the entire operation will fail.

<details>
<summary>🔨 See the suggested fix</summary>

``` diff
@@ -430,41 +430,47 @@
         return [];
       }
 
-      // Handle owner information
-      if (includeOwner || forRoundRobinSkip) {
-        const ownerIds: Set<string> = new Set();
-
-        if (accountOwnerId) {
-          ownerIds.add(accountOwnerId);
-        } else {
-          records.forEach((record) => {
-            if (record?.OwnerId) {
-              ownerIds.add(record.OwnerId);
-            }
+        // Handle owner information
+        if (includeOwner || forRoundRobinSkip) {
+          const ownerIds: Set<string> = new Set();
+
+          if (accountOwnerId) {
+            ownerIds.add(accountOwnerId);
+          } else {
+            records.forEach((record) => {
+              if (record?.OwnerId) {
+                ownerIds.add(record.OwnerId);
+              }
+            });
+          }
+
+          const ownersQuery =
+            ownerIds.size > 0
+              ? await Promise.all(
+                  Array.from(ownerIds).map(async (ownerId) => {
+                    try {
+                      const result = await this.getSalesforceUserFromUserId(ownerId);
+                      return result;
+                    } catch (error) {
+                      log.error("Error fetching owner information", { ownerId, error });
+                      return null;
+                    }
+                  })
+                )
+              : [];
+
+          // Filter out any undefined results and ensure records exist
+          const validOwnersQuery = ownersQuery.filter((query): query is jsforce.QueryResult<ContactRecord> => {
+            if (!query) return false;
+            const firstRecord = query?.records?.[0];
+            if (!firstRecord) return false;
+
+            return (
+              typeof firstRecord === "object" && "Email" in firstRecord && typeof firstRecord.Email === "string"
+            );
           });
-        }
-
-        const ownersQuery =
-          ownerIds.size > 0
-            ? await Promise.all(
-                Array.from(ownerIds).map(async (ownerId) => {
-                  const result = await this.getSalesforceUserFromUserId(ownerId);
-                  return result;
-                })
-              )
-            : [];
-
-        // Filter out any undefined results and ensure records exist
-        const validOwnersQuery = ownersQuery.filter((query): query is jsforce.QueryResult<ContactRecord> => {
-          const firstRecord = query?.records?.[0];
-          if (!firstRecord) return false;
-
-          return (
-            typeof firstRecord === "object" && "Email" in firstRecord && typeof firstRecord.Email === "string"
-          );
-        });
-
-        return records
+
+          return records
           .map((record) => {
             if (!record?.Id) {
               log.warn("Invalid record data", { record });

```

[Click Here](http://localhost:3006/commit/7332235d-c213-4641-91ca-f18bfcfa395d)
<sub>⚠️ Fume is an LLM-based tool and can make mistakes.</sub>
</details>

File: packages/app-store/salesforce/lib/CrmService.test.ts [hunk 3]
```diff
@@ -41,6 +100,11 @@ describe("SalesforceCRMService", () => {
     service.conn = Promise.resolve(mockConnection);
   });
 
+  const mockAppOptions = (appOptions: AppOptions) => {
```
Comment (fumeoss) [inline packages/app-store/salesforce/lib/CrmService.test.ts:103]: Consider adding test cases for when appDataSchema fields are undefined. The schema allows optional fields but we're not testing those scenarios.

File: packages/app-store/salesforce/lib/CrmService.test.ts (hunk 1 above)
Comment (fumeoss) [inline packages/app-store/salesforce/lib/CrmService.test.ts:14]: Consider moving mock response objects into a separate test fixtures file to improve test readability and maintainability.

File: packages/app-store/salesforce/lib/CrmService.test.ts (hunk 3 above)
Comment (fumeoss) [inline packages/app-store/salesforce/lib/CrmService.test.ts:103]: Move mockAppOptions helper function to beforeEach to avoid repetitive setup code in each test.

<details>
<summary>🔨 See the suggested fix</summary>

``` diff
@@ -98,8 +98,15 @@
     // eslint-disable-next-line @typescript-eslint/ban-ts-comment
     // @ts-ignore
     service.conn = Promise.resolve(mockConnection);
+    
+    global.mockAppOptions = (appOptions: AppOptions) => {
+      const appOptionsSpy = vi.spyOn(service, "getAppOptions");
+      appOptionsSpy.mockReturnValue(appOptions);
+    };
   });
 
+  describe("getContacts", () => {
+    it("should return empty array when no records found", async () => {
   const mockAppOptions = (appOptions: AppOptions) => {
     const appOptionsSpy = vi.spyOn(service, "getAppOptions");
     appOptionsSpy.mockReturnValue(appOptions);

```

[Click Here](http://localhost:3006/commit/e64e598e-0bf6-4a01-a402-2f311daaf600)
<sub>⚠️ Fume is an LLM-based tool and can make mistakes.</sub>
</details>

File: packages/app-store/salesforce/lib/CrmService.test.ts [hunk 4]
```diff
@@ -130,26 +194,25 @@ describe("SalesforceCRMService", () => {
...
     });
 
-    it("should include owner information when includeOwner is true", async () => {
+    it("should return contacts when includeOwner is true but no owner is found", async () => {
+      mockAppOptions({
+        roundRobinSkipCheckRecordOn: SalesforceRecordEnum.CONTACT,
+        createEventOn: SalesforceRecordEnum.CONTACT,
+      });
+
       const contactQueryResponse = {
```
Comment (fumeoss) [inline packages/app-store/salesforce/lib/CrmService.test.ts:203]: Consider flattening the deeply nested describe blocks by combining them or using more descriptive test names.

File: packages/app-store/salesforce/lib/CrmService.ts [hunk 5]
```diff
@@ -444,17 +444,15 @@ export default class SalesforceCRMService implements CRM {
           });
```
Comment (fumeoss) [inline packages/app-store/salesforce/lib/CrmService.ts:444]: Removing the early return here could cause unexpected behavior in integrations that expect empty results when no owners are found. Consider keeping the warning but make the early return conditional on whether includeOwner is true.

<details>
<summary>🔨 See the suggested fix</summary>

``` diff
@@ -457,7 +457,10 @@
         // Filter out any undefined results and ensure records exist
         const validOwnersQuery = ownersQuery.filter((query): query is jsforce.QueryResult<ContactRecord> => {
           const firstRecord = query?.records?.[0];
-          if (!firstRecord) return false;
+          if (!firstRecord) {
+            log.warn("No owner record found for query");
+            return false;
+          }
 
           return (
             typeof firstRecord === "object" && "Email" in firstRecord && typeof firstRecord.Email === "string"

```

[Click Here](http://localhost:3006/commit/025f05b7-4a6f-494b-a0fe-e9380cf6660c)
<sub>⚠️ Fume is an LLM-based tool and can make mistakes.</sub>
</details>

File: packages/app-store/salesforce/lib/CrmService.test.ts (hunk 1 above)
Comment (fumeoss) [inline packages/app-store/salesforce/lib/CrmService.test.ts:14]: The test data objects should be moved inside the describe block to follow encapsulation pattern. This prevents potential test data mutations affecting other tests.

**Comment by fumeoss** [comment]
## Summary
- This PR contains unit tests for the SalesforceCRMService, specifically focusing on the `getContacts` method
- Tests cover various scenarios including handling single and multiple email inputs, returning empty results, including owner information, and handling different record types
- The test suite uses mock connections and spies to verify Salesforce query behavior, testing edge cases like round-robin skip and owner information retrieval

:mag: **Fume is reviewing this PR!**

:link: Track the review progress here:
<a href="http://localhost:3006/chat/bc1a3340-1aaa-41f0-994a-0bdceea64056" target="_blank">http://localhost:3006/chat/bc1a3340-1aaa-41f0-994a-0bdceea64056</a>

File: packages/app-store/salesforce/lib/CrmService.test.ts [hunk 6]
```diff
@@ -1,12 +1,71 @@
...
+      OwnerId: "owner001",
+      Email: "test@example.com",
+      attributes: { type: "Account" },
+    },
+  ],
+};
+
 describe("SalesforceCRMService", () => {
   let service: SalesforceCRMService;
   let mockConnection: { query: any };
```
Comment (fumeoss) [inline packages/app-store/salesforce/lib/CrmService.test.ts:71]: Move mockCredential setup into each test case to prevent test pollution. Some tests override configuration which could affect other tests.

<details>
<summary>🔨 See the suggested fix</summary>

``` diff
@@ -72,45 +72,63 @@
 
   setupAndTeardown();
 
-  beforeEach(() => {
-    mockConnection = {
-      query: vi.fn(),
-    };
-
-    const mockCredential: CredentialPayload = {
-      id: 1,
-      type: "salesforce_crm",
-      key: {
-        access_token: "mock_token",
-        refresh_token: "mock_refresh",
-        instance_url: "https://mock.salesforce.com",
-      },
-      userId: 1,
-      appId: "salesforce",
-      teamId: null,
-      invalid: false,
-      user: {
-        email: "test-user@example.com",
-      },
-    };
-
-    service = new SalesforceCRMService(mockCredential, {}, true);
-    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
-    // @ts-ignore
-    service.conn = Promise.resolve(mockConnection);
-  });
-
-  const mockAppOptions = (appOptions: AppOptions) => {
-    const appOptionsSpy = vi.spyOn(service, "getAppOptions");
-    appOptionsSpy.mockReturnValue(appOptions);
-  };
-
   describe("getContacts", () => {
     it("should return empty array when no records found", async () => {
+      const mockCredential: CredentialPayload = {
+        id: 1,
+        type: "salesforce_crm",
+        key: {
+          access_token: "mock_token",
+          refresh_token: "mock_refresh",
+          instance_url: "https://mock.salesforce.com",
+        },
+        userId: 1,
+        appId: "salesforce",
+        teamId: null,
+        invalid: false,
+        user: {
+          email: "test-user@example.com",
+        },
+      };
+
+      service = new SalesforceCRMService(mockCredential, {}, true);
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore
+      service.conn = Promise.resolve(mockConnection);
+
       // Setup mock response
       mockConnection.query.mockResolvedValueOnce({ records: [] });
 
       const result = await service.getContacts({ emails: "test@example.com" });
+
+      expect(result).toEqual([]);
+      expect(mockConnection.query).toHaveBeenCalledWith(
+        "SELECT Id, Email, OwnerId FROM Contact WHERE Email IN ('test@example.com')"
+      );
+    });
+
+    it("should handle single email string input", async () => {
+      const mockCredential: CredentialPayload = {
+        id: 1,
+        type: "salesforce_crm",
+        key: {
+          access_token: "mock_token",
+          refresh_token: "mock_refresh",
+          instance_url: "https://mock.salesforce.com",
+        },
+        userId: 1,
+        appId: "salesforce",
+        teamId: null,
+        invalid: false,
+        user: {
+          email: "test-user@example.com",
+        },
+      };
+
+      service = new SalesforceCRMService(mockCredential, {}, true);
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore
+      service.conn = Promise.resolve(mockConnection);
 
       expect(result).toEqual([]);
       expect(mockConnection.query).toHaveBeenCalledWith(

```

[Click Here](http://localhost:3006/commit/a08a2b55-15d4-4e9d-a63b-4b3bb41f450f)
<sub>⚠️ Fume is an LLM-based tool and can make mistakes.</sub>
</details>

File: packages/app-store/salesforce/lib/CrmService.test.ts [hunk 7]
```diff
@@ -177,89 +240,297 @@ describe("SalesforceCRMService", () => {
...
+          createEventOn: SalesforceRecordEnum.LEAD,
+        });
+
+        const querySpy = vi.spyOn(mockConnection, "query");
+        querySpy.mockResolvedValueOnce(leadQueryResponse);
+
+        const result = await service.getContacts({
+          emails: "test@example.com",
+        });
+
```
Comment (fumeoss) [inline packages/app-store/salesforce/lib/CrmService.test.ts:282]: This test should verify that a warning is logged when owner information is missing. Add log verification to ensure proper error handling.

<details>
<summary>🔨 See the suggested fix</summary>

``` diff
@@ -215,6 +215,8 @@
         records: [],
       };
 
+      const consoleSpy = vi.spyOn(console, "warn");
+
       // Setup query spy with specific responses
       const querySpy = vi.spyOn(mockConnection, "query");
       querySpy
@@ -236,6 +238,7 @@
         },
       ]);
 
+      expect(consoleSpy).toHaveBeenCalledWith("No owner found for contact with ID: 001");
       expect(querySpy).toHaveBeenNthCalledWith(
         1,
         "SELECT Id, Email, OwnerId FROM Contact WHERE Email IN ('test@example.com')"

```

[Click Here](http://localhost:3006/commit/cde856c3-f3b8-4642-9ce7-b071f5cebebc)
<sub>⚠️ Fume is an LLM-based tool and can make mistakes.</sub>
</details>

File: packages/app-store/salesforce/lib/CrmService.ts (hunk 5 above)
Comment (fumeoss) [inline packages/app-store/salesforce/lib/CrmService.ts:444]: Add warning logs when owner information is missing to help with debugging. Completely silent failures could hide real issues.

<details>
<summary>🔨 See the suggested fix</summary>

``` diff
@@ -440,6 +440,8 @@
           records.forEach((record) => {
             if (record?.OwnerId) {
               ownerIds.add(record.OwnerId);
+            } else {
+              log.warn("Record missing ownerId", safeStringify({ record }));
             }
           });
         }
@@ -449,6 +451,9 @@
             ? await Promise.all(
                 Array.from(ownerIds).map(async (ownerId) => {
                   const result = await this.getSalesforceUserFromUserId(ownerId);
+                  if (!result?.records?.length) {
+                    log.warn("Owner information not found", safeStringify({ ownerId }));
+                  }
                   return result;
                 })
               )

```

[Click Here](http://localhost:3006/commit/a2fb4296-3644-4081-bb62-8fadfa3059c6)
<sub>⚠️ Fume is an LLM-based tool and can make mistakes.</sub>
</details>

**Comment by fumeoss** [comment]
### Testing
- Extended existing CrmService.test.ts with edge cases for owner handling
- Tested contact without owner ID - verified basic contact data returns correctly
- Tested contact with invalid owner ID - verified partial data returns without error
- Verified existing tests covered multiple contacts, owner information inclusion, and account record types
- No bugs found - all test cases passed and verified the intended flexible owner handling behavior
- Missing error logging for owner lookup failures could make debugging harder in production (noted in code review)

The tests confirm the changes work as intended but could benefit from better error logging for troubleshooting.

👋 This PR was tested by [Fume](https://fumedev.com) - an AI agent that can use an isolated development environment to test and review your PRs. We would love to make Fume a regular reviewer at `emregucerr`! (it's free for open-source repos) If that sounds interesting, [let's chat](https://cal.com/fumedev/30min).
If you want us to stop reviewing your PRs, click [here](mailto:contact@fumedev.com?subject=Hide%20My%20Open-Source%20Project%20From%20Fume&body=Stop:%20emregucerr).
